apps/interpret_nn/delegate/halide_delegate.cpp

```diff
@@ -0,0 +1,881 @@
+        auto input = GetTensorById(context, node->inputs->data[0]);
+        auto output = GetTensorById(context, node->outputs->data[0]);
+        const TfLiteReshapeParams *params = (const TfLiteReshapeParams *)(node->builtin_data);
+        std::vector<int> new_shape;
+        new_shape.assign(params->shape, params->shape + params->num_dimensions);
+        return make_unique<ReshapeOp>(input, output, new_shape);
+    }
+
+    std::unique_ptr<Op> BuildQuantize(TfLiteContext *context, TfLiteNode *node) {
+        auto input = GetTensorById(context, node->inputs->data[0]);
+        auto output = GetTensorById(context, node->outputs->data[0]);
+        return make_unique<QuantizeOp>(input, output);
+    }
+
+    const HalideDelegateOptions options_;
+    std::unique_ptr<Model> model_;
+    std::unique_ptr<ModelInterpreter> interpreter_;
+    std::map<int, std::shared_ptr<Tensor>> tensor_id_to_tensor_ptr_;
+};
+
+bool InputsHaveCorrectTypes(const TfLiteNode *node,
+                            TfLiteContext *context,
+                            std::initializer_list<int> per_input_possible_types_mask) {
+    if (node->inputs->size != (int)per_input_possible_types_mask.size()) {
+        LOG(ERROR) << "inputs size mismatch in InputsHaveCorrectTypes";
+        return false;
+    }
+    int i = -1;
+    for (int possible_types_mask : per_input_possible_types_mask) {
+        ++i;
+        // Skip optional tensor.
+        const int tensor_id = node->inputs->data[i];
+        if (tensor_id == kTfLiteOptionalTensor) {
+            continue;
+        }
+        const TfLiteTensor &tensor = context->tensors[tensor_id];
+        const int tensor_type_mask = 1 << tensor.type;
+        if (!(tensor_type_mask & possible_types_mask)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool AllInputsHaveType(const TfLiteNode *node, TfLiteContext *context, int possible_types_mask) {
+    for (int i = 0; i < node->inputs->size; ++i) {
+        const int tensor_id = node->inputs->data[i];
+        if (tensor_id == kTfLiteOptionalTensor) {
+            continue;
+        }
+        const TfLiteTensor &tensor = context->tensors[tensor_id];
+        const int tensor_type_mask = 1 << tensor.type;
+        if (!(tensor_type_mask & possible_types_mask)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool IsActivationReluOrNone(TfLiteFusedActivation activation) {
+    return (activation == kTfLiteActRelu ||
+            activation == kTfLiteActRelu6 ||
+            activation == kTfLiteActReluN1To1 ||
+            activation == kTfLiteActNone);
+}
+
+// TODO: this should also allow Int8 once we fix biasing for those
+constexpr int k8BitMask = 1 << kTfLiteUInt8;
+
+bool IsNodeSupported_Add(TfLiteContext *context, TfLiteNode *node, TfLiteRegistration *registration) {
+    if (!(registration->version <= 2)) {
+        return false;
+    }
+    if (!InputsHaveCorrectTypes(
+            node, context,
```

Comment:
Super nit, but do you want these extra line breaks here? IMO when modifying code like this, it's a bit annoying if your text editor doesn't do the right thing. And this isn't a very long line.

There are other instances of this, but this one is the most stark (2 extra line breaks on a pretty short line).

Reply:
Nah, this is a case of me being lazy and relying on clang-format and it deciding this is ok, apparently. Fixing.